nemo/collections/tts/models/fastpitch_ssl.py

```diff
@@ -0,0 +1,504 @@
+            if isinstance(self.logger, LoggerCollection):
+                for logger in self.logger:
+                    if isinstance(logger, TensorBoardLogger):
+                        tb_logger = logger.experiment
+                        break
+            self._tb_logger = tb_logger
+        return self._tb_logger
+
+    def forward(
+        self,
+        *,
+        text,
+        durs=None,
+        pitch=None,
+        speaker=None,
+        pace=1.0,
+        spec=None,
+        attn_prior=None,
+        mel_lens=None,
+        input_lens=None,
+        enc_out=None,
+        enc_mask=None,
+    ):
+        return self.fastpitch(
+            text=text,
+            durs=durs,
+            pitch=pitch,
+            speaker=speaker,
+            pace=pace,
+            spec=spec,
+            attn_prior=attn_prior,
+            mel_lens=mel_lens,
+            input_lens=input_lens,
+            enc_out=enc_out,
+            enc_mask=enc_mask,
+        )
+
+    @typecheck(output_types={"spect": NeuralType(('B', 'D', 'T_spec'), MelSpectrogramType())})
+    def generate_spectrogram(
+        self, tokens: 'torch.tensor', speaker: Optional[int] = None, pace: float = 1.0
+    ) -> torch.tensor:
+        if self.training:
+            logging.warning("generate_spectrogram() is meant to be called in eval mode.")
+        if isinstance(speaker, int):
+            speaker = torch.tensor([speaker]).to(self.device)
+        spect, *_ = self(text=tokens, durs=None, pitch=None, speaker=speaker, pace=pace)
+        return spect
+
+    def compute_encoding(self, content_embedding, speaker_embedding, dataset_id=None):
+        # content embedding is (B, C, T)
+        # speaker embedding is (B, C)
+        # pitch_contour is (B, T)
+        content_embedding = content_embedding.permute(0, 2, 1)  # (B, C, T) -> (B, T, C)
+        content_embedding_projected = self.content_projection_layer(content_embedding)
+        content_embedding_projected = content_embedding_projected.permute(0, 2, 1)
+        speaker_embedding_projected = self.speaker_projection_layer(speaker_embedding)
+        speaker_embedding_repeated = speaker_embedding_projected[:, :, None].repeat(
+            1, 1, content_embedding_projected.shape[2]
+        )
+
+        encoded = torch.cat([content_embedding_projected, speaker_embedding_repeated], dim=1)
+
+        encoded = encoded.permute(0, 2, 1)  # (B, C, T) -> (B, T, C)
+
+        if self.num_datasets > 1:
+            dataset_embedding = self.dataset_embedding_layer(dataset_id)  # (B, C)
+            dataset_embedding_repeated = dataset_embedding[:, None, :].repeat(1, encoded.shape[1], 1)
+            encoded = encoded + dataset_embedding_repeated
+
+        return encoded
+
+    def compute_speaker_loss(self, generated_mel, gt_speaker_embedding):
+        ssl_model = self.non_trainable_models['ssl_model']
+        ssl_model.to(self.device)
+
+        generated_mel_sliced = generated_mel[:, :, :172]
```

Comment:
This was for using 2 second utterances. Added a comment about it. We are not actually using this loss in our final checkpoints since this did not help. Can drop this functionality if you want (or mark it as experimental/TBD ?)

Reply:
dropped this